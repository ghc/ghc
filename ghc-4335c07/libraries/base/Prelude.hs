{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Prelude
-- Copyright   :  (c) The University of Glasgow 2001
-- License     :  BSD-style (see the file libraries/base/LICENSE)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  stable
-- Portability :  portable
--
-- The Prelude: a standard module. The Prelude is imported by default
-- into all Haskell modules unless either there is an explicit import
-- statement for it, or the NoImplicitPrelude extension is enabled.
--
-----------------------------------------------------------------------------

module Prelude (

    -- * Standard types, classes and related functions

    -- ** Basic data types
    Bool(False, True),
    (&&), (||), not, otherwise,

    Maybe(Nothing, Just),
    maybe,

    Either(Left, Right),
    either,

    Ordering(LT, EQ, GT),
    Char, String,

    -- *** Tuples
    fst, snd, curry, uncurry,

    -- ** Basic type classes
    Eq((==), (/=)),
    Ord(compare, (<), (<=), (>=), (>), max, min),
    Enum(succ, pred, toEnum, fromEnum, enumFrom, enumFromThen,
         enumFromTo, enumFromThenTo),
    Bounded(minBound, maxBound),

    -- ** Numbers

    -- *** Numeric types
    Int, Integer, Float, Double,
    Rational, Word,

    -- *** Numeric type classes
    Num((+), (-), (*), negate, abs, signum, fromInteger),
    Real(toRational),
    Integral(quot, rem, div, mod, quotRem, divMod, toInteger),
    Fractional((/), recip, fromRational),
    Floating(pi, exp, log, sqrt, (**), logBase, sin, cos, tan,
             asin, acos, atan, sinh, cosh, tanh, asinh, acosh, atanh),
    RealFrac(properFraction, truncate, round, ceiling, floor),
    RealFloat(floatRadix, floatDigits, floatRange, decodeFloat,
              encodeFloat, exponent, significand, scaleFloat, isNaN,
              isInfinite, isDenormalized, isIEEE, isNegativeZero, atan2),

    -- *** Numeric functions
    subtract, even, odd, gcd, lcm, (^), (^^),
    fromIntegral, realToFrac,

    -- ** Semigroups and Monoids
    Semigroup((<>)),
    Monoid(mempty, mappend, mconcat),

    -- ** Monads and functors
    Functor(fmap, (<$)), (<$>),
    Applicative(pure, (<*>), (*>), (<*)),
    Monad((>>=), (>>), return, fail),
    mapM_, sequence_, (=<<),

    -- ** Folds and traversals
    Foldable(elem,      -- :: (Foldable t, Eq a) => a -> t a -> Bool
             -- fold,   -- :: Monoid m => t m -> m
             foldMap,   -- :: Monoid m => (a -> m) -> t a -> m
             foldr,     -- :: (a -> b -> b) -> b -> t a -> b
             -- foldr', -- :: (a -> b -> b) -> b -> t a -> b
             foldl,     -- :: (b -> a -> b) -> b -> t a -> b
             -- foldl', -- :: (b -> a -> b) -> b -> t a -> b
             foldr1,    -- :: (a -> a -> a) -> t a -> a
             foldl1,    -- :: (a -> a -> a) -> t a -> a
             maximum,   -- :: (Foldable t, Ord a) => t a -> a
             minimum,   -- :: (Foldable t, Ord a) => t a -> a
             product,   -- :: (Foldable t, Num a) => t a -> a
             sum),      -- :: Num a => t a -> a
             -- toList) -- :: Foldable t => t a -> [a]

    Traversable(traverse, sequenceA, mapM, sequence),

    -- ** Miscellaneous functions
    id, const, (.), flip, ($), until,
    asTypeOf, error, errorWithoutStackTrace, undefined,
    seq, ($!),

    -- * List operations
    map, (++), filter,
    head, last, tail, init, null, length, (!!),
    reverse,
    -- *** Special folds
    and, or, any, all,
    concat, concatMap,
    -- ** Building lists
    -- *** Scans
    scanl, scanl1, scanr, scanr1,
    -- *** Infinite lists
    iterate, repeat, replicate, cycle,
    -- ** Sublists
    take, drop, splitAt, takeWhile, dropWhile, span, break,
    -- ** Searching lists
    notElem, lookup,
    -- ** Zipping and unzipping lists
    zip, zip3, zipWith, zipWith3, unzip, unzip3,
    -- ** Functions on strings
    lines, words, unlines, unwords,

    -- * Converting to and from @String@
    -- ** Converting to @String@
    ShowS,
    Show(showsPrec, showList, show),
    shows,
    showChar, showString, showParen,
    -- ** Converting from @String@
    ReadS,
    Read(readsPrec, readList),
    reads, readParen, read, lex,

    -- * Basic Input and output
    IO,
    -- ** Simple I\/O operations
    -- All I/O functions defined here are character oriented.  The
    -- treatment of the newline character will vary on different systems.
    -- For example, two characters of input, return and linefeed, may
    -- read as a single newline character.  These functions cannot be
    -- used portably for binary I/O.
    -- *** Output functions
    putChar,
    putStr, putStrLn, print,
    -- *** Input functions
    getChar,
    getLine, getContents, interact,
    -- *** Files
    FilePath,
    readFile, writeFile, appendFile, readIO, readLn,
    -- ** Exception handling in the I\/O monad
    IOError, ioError, userError,

  ) where

import Control.Monad
import System.IO
import System.IO.Error
import Data.List
import Data.Either
import Data.Foldable    ( Foldable(..) )
import Data.Functor     ( (<$>) )
import Data.Maybe
import Data.Traversable ( Traversable(..) )
import Data.Tuple

import GHC.Base hiding ( foldr, mapM, sequence )
import Text.Read
import GHC.Enum
import GHC.Num
import GHC.Real
import GHC.Float
import GHC.Show
