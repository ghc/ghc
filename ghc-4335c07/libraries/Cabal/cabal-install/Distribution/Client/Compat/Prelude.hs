{-# LANGUAGE CPP #-}

-- to suppress WARNING in "Distribution.Compat.Prelude.Internal"
{-# OPTIONS_GHC -fno-warn-deprecations #-}

-- | This module does two things:
--
-- * Acts as a compatiblity layer, like @base-compat@.
--
-- * Provides commonly used imports.
--
-- This module is a superset of "Distribution.Compat.Prelude" (which
-- this module re-exports)
--
module Distribution.Client.Compat.Prelude
  ( module Distribution.Compat.Prelude.Internal
  , Prelude.IO
  , readMaybe
  ) where

import Prelude (IO)
import Distribution.Compat.Prelude.Internal hiding (IO)

#if MIN_VERSION_base(4,6,0)
import Text.Read
         ( readMaybe )
#endif

#if !MIN_VERSION_base(4,6,0)
-- | An implementation of readMaybe, for compatibility with older base versions.
readMaybe :: Read a => String -> Maybe a
readMaybe s = case reads s of
                [(x,"")] -> Just x
                _        -> Nothing
#endif
