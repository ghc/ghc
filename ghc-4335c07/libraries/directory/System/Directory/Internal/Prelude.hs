{-# LANGUAGE CPP #-}
-- |
-- Stability: unstable
-- Portability: portable
--
-- Internal modules are always subject to change from version to version.

module System.Directory.Internal.Prelude
  ( module Prelude
#if MIN_VERSION_base(4, 8, 0)
  , module Data.Void
#else
  , module Control.Applicative
  , module Data.Functor
  , Void
#endif
  , module Control.Arrow
  , module Control.Concurrent
  , module Control.Exception
  , module Control.Monad
  , module Data.Bits
  , module Data.Char
  , module Data.Foldable
  , module Data.Function
  , module Data.Maybe
  , module Data.Monoid
  , module Data.IORef
  , module Data.Traversable
  , module Foreign
  , module Foreign.C
  , module GHC.IO.Encoding
  , module GHC.IO.Exception
  , module System.Environment
  , module System.Exit
  , module System.IO
  , module System.IO.Error
  , module System.Posix.Internals
  , module System.Posix.Types
  , module System.Timeout
  ) where
#if !MIN_VERSION_base(4, 6, 0)
import Prelude hiding (catch)
#endif
#if MIN_VERSION_base(4, 8, 0)
import Data.Void (Void)
#else
import Control.Applicative ((<*>), pure)
import Data.Functor ((<$>), (<$))
#endif
import Control.Arrow (second)
import Control.Concurrent
  ( forkIO
  , killThread
  , newEmptyMVar
  , putMVar
  , readMVar
  , takeMVar
  )
import Control.Exception
  ( SomeException
  , bracket
  , bracket_
  , catch
  , finally
  , mask
  , onException
  , throwIO
  , try
  )
import Control.Monad ((>=>), (<=<), unless, when, replicateM_)
import Data.Bits ((.&.), (.|.), complement)
import Data.Char (isAlpha, isAscii, toLower, toUpper)
import Data.Foldable (for_, traverse_)
import Data.Function (on)
import Data.Maybe (catMaybes, fromMaybe, maybeToList)
import Data.Monoid ((<>), mconcat, mempty)
import Data.IORef (IORef, newIORef, readIORef, writeIORef)
import Data.Traversable (for)
import Foreign
  ( Ptr
  , Storable
    ( alignment
    , peek
    , peekByteOff
    , peekElemOff
    , poke
    , pokeByteOff
    , pokeElemOff
    , sizeOf
    )
  , alloca
  , allocaArray
  , allocaBytes
  , allocaBytesAligned
  , maybeWith
  , nullPtr
  , plusPtr
  , with
  , withArray
  )
import Foreign.C
  ( CInt(..)
  , CLong(..)
  , CString
  , CTime(..)
  , CUChar(..)
  , CULong(..)
  , CUShort(..)
  , CWString
  , CWchar(..)
  , peekCString
  , peekCWStringLen
  , throwErrnoIfMinus1Retry_
  , throwErrnoIfMinus1_
  , throwErrnoIfNull
  , throwErrnoPathIfMinus1_
  , withCString
  , withCWString
  )
import GHC.IO.Exception
  ( IOErrorType
    ( InappropriateType
    , OtherError
    , UnsupportedOperation
    )
  )
import GHC.IO.Encoding (getFileSystemEncoding)
import System.Environment (getArgs, getEnv)
import System.Exit (exitFailure)
import System.IO
  ( Handle
  , IOMode(ReadMode, WriteMode)
  , hClose
  , hFlush
  , hGetBuf
  , hPutBuf
  , hPutStr
  , hPutStrLn
  , openBinaryTempFile
  , stderr
  , stdout
  , withBinaryFile
  )
import System.IO.Error
  ( IOError
  , catchIOError
  , illegalOperationErrorType
  , ioeGetErrorString
  , ioeGetErrorType
  , ioeGetLocation
  , ioeSetErrorString
  , ioeSetFileName
  , ioeSetLocation
  , isAlreadyExistsError
  , isDoesNotExistError
  , isIllegalOperation
  , isPermissionError
  , mkIOError
  , modifyIOError
  , permissionErrorType
  , tryIOError
  , userError
  )
import System.Posix.Internals (withFilePath)
import System.Posix.Types (EpochTime)
import System.Timeout (timeout)

#if !MIN_VERSION_base(4, 8, 0)
data Void = Void

_unusedVoid :: Void
_unusedVoid = Void
#endif
