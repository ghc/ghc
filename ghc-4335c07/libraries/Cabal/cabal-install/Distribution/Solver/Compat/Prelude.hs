-- to suppress WARNING in "Distribution.Compat.Prelude.Internal"
{-# OPTIONS_GHC -fno-warn-deprecations #-}

-- | This module does two things:
--
-- * Acts as a compatiblity layer, like @base-compat@.
--
-- * Provides commonly used imports.
--
-- This module is a superset of "Distribution.Compat.Prelude" (which
-- this module re-exports)
--
module Distribution.Solver.Compat.Prelude
  ( module Distribution.Compat.Prelude.Internal
  , Prelude.IO
  ) where

import Prelude (IO)
import Distribution.Compat.Prelude.Internal hiding (IO)
