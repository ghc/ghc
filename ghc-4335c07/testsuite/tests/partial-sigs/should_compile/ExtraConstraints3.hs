{-# LANGUAGE PartialTypeSignatures #-}
{-# LANGUAGE NoMonomorphismRestriction #-}
module ExtraConstraints3 where

import qualified Prelude as P

import Prelude (Bool, Bounded, Char, Either, Enum, Eq, FilePath, Floating,
                Fractional, Functor, IO, IOError, Int, Integer, Integral,
                Maybe, Monad, Num, Ord, Ordering, Rational, Read, ReadS, Real,
                RealFloat, RealFrac, Show, ShowS, String)

-- Proof by enumeration! jk :p
-- All of Prelude typechecks given the dummy type signature `_ => _`,
-- which is the same as omitting the type signature entirely.

(!!) :: _ => _
(!!) = (P.!!)
($!) :: _ => _
($!) = (P.$!)
($) :: _ => _
($) = (P.$)
(&&) :: _ => _
(&&) = (P.&&)
(*) :: _ => _
(*) = (P.*)
(**) :: _ => _
(**) = (P.**)
(+) :: _ => _
(+) = (P.+)
(++) :: _ => _
(++) = (P.++)
(-) :: _ => _
(-) = (P.-)
(.) :: _ => _
(.) = (P..)
(/) :: _ => _
(/) = (P./)
(/=) :: _ => _
(/=) = (P./=)
(<) :: _ => _
(<) = (P.<)
(<=) :: _ => _
(<=) = (P.<=)
(=<<) :: _ => _
(=<<) = (P.=<<)
(==) :: _ => _
(==) = (P.==)
(>) :: _ => _
(>) = (P.>)
(>=) :: _ => _
(>=) = (P.>=)
(>>) :: _ => _
(>>) = (P.>>)
(>>=) :: _ => _
(>>=) = (P.>>=)
(^) :: _ => _
(^) = (P.^)
(^^) :: _ => _
(^^) = (P.^^)
(||) :: _ => _
(||) = (P.||)
abs :: _ => _
abs = P.abs
acos :: _ => _
acos = P.acos
acosh :: _ => _
acosh = P.acosh
all :: _ => _
all = P.all
and :: _ => _
and = P.and
any :: _ => _
any = P.any
appendFile :: _ => _
appendFile = P.appendFile
asTypeOf :: _ => _
asTypeOf = P.asTypeOf
asin :: _ => _
asin = P.asin
asinh :: _ => _
asinh = P.asinh
atan :: _ => _
atan = P.atan
atan2 :: _ => _
atan2 = P.atan2
atanh :: _ => _
atanh = P.atanh
break :: _ => _
break = P.break
ceiling :: _ => _
ceiling = P.ceiling
compare :: _ => _
compare = P.compare
concat :: _ => _
concat = P.concat
concatMap :: _ => _
concatMap = P.concatMap
const :: _ => _
const = P.const
cos :: _ => _
cos = P.cos
cosh :: _ => _
cosh = P.cosh
curry :: _ => _
curry = P.curry
cycle :: _ => _
cycle = P.cycle
decodeFloat :: _ => _
decodeFloat = P.decodeFloat
div :: _ => _
div = P.div
divMod :: _ => _
divMod = P.divMod
drop :: _ => _
drop = P.drop
dropWhile :: _ => _
dropWhile = P.dropWhile
either :: _ => _
either = P.either
elem :: _ => _
elem = P.elem
encodeFloat :: _ => _
encodeFloat = P.encodeFloat
enumFrom :: _ => _
enumFrom = P.enumFrom
enumFromThen :: _ => _
enumFromThen = P.enumFromThen
enumFromThenTo :: _ => _
enumFromThenTo = P.enumFromThenTo
enumFromTo :: _ => _
enumFromTo = P.enumFromTo
error :: _ => _
error = P.error
even :: _ => _
even = P.even
exp :: _ => _
exp = P.exp
exponent :: _ => _
exponent = P.exponent
fail :: _ => _
fail = P.fail
filter :: _ => _
filter = P.filter
flip :: _ => _
flip = P.flip
floatDigits :: _ => _
floatDigits = P.floatDigits
floatRadix :: _ => _
floatRadix = P.floatRadix
floatRange :: _ => _
floatRange = P.floatRange
floor :: _ => _
floor = P.floor
fmap :: _ => _
fmap = P.fmap
foldl :: _ => _
foldl = P.foldl
foldl1 :: _ => _
foldl1 = P.foldl1
foldr :: _ => _
foldr = P.foldr
foldr1 :: _ => _
foldr1 = P.foldr1
fromEnum :: _ => _
fromEnum = P.fromEnum
fromInteger :: _ => _
fromInteger = P.fromInteger
fromIntegral :: _ => _
fromIntegral = P.fromIntegral
fromRational :: _ => _
fromRational = P.fromRational
fst :: _ => _
fst = P.fst
gcd :: _ => _
gcd = P.gcd
getChar :: _ => _
getChar = P.getChar
getContents :: _ => _
getContents = P.getContents
getLine :: _ => _
getLine = P.getLine
head :: _ => _
head = P.head
id :: _ => _
id = P.id
init :: _ => _
init = P.init
interact :: _ => _
interact = P.interact
ioError :: _ => _
ioError = P.ioError
isDenormalized :: _ => _
isDenormalized = P.isDenormalized
isIEEE :: _ => _
isIEEE = P.isIEEE
isInfinite :: _ => _
isInfinite = P.isInfinite
isNaN :: _ => _
isNaN = P.isNaN
isNegativeZero :: _ => _
isNegativeZero = P.isNegativeZero
iterate :: _ => _
iterate = P.iterate
last :: _ => _
last = P.last
lcm :: _ => _
lcm = P.lcm
length :: _ => _
length = P.length
lex :: _ => _
lex = P.lex
lines :: _ => _
lines = P.lines
log :: _ => _
log = P.log
logBase :: _ => _
logBase = P.logBase
lookup :: _ => _
lookup = P.lookup
map :: _ => _
map = P.map
mapM :: _ => _
mapM = P.mapM
mapM_ :: _ => _
mapM_ = P.mapM_
max :: _ => _
max = P.max
maxBound :: _ => _
maxBound = P.maxBound
maximum :: _ => _
maximum = P.maximum
maybe :: _ => _
maybe = P.maybe
min :: _ => _
min = P.min
minBound :: _ => _
minBound = P.minBound
minimum :: _ => _
minimum = P.minimum
mod :: _ => _
mod = P.mod
negate :: _ => _
negate = P.negate
not :: _ => _
not = P.not
notElem :: _ => _
notElem = P.notElem
null :: _ => _
null = P.null
odd :: _ => _
odd = P.odd
or :: _ => _
or = P.or
otherwise :: _ => _
otherwise = P.otherwise
pi :: _ => _
pi = P.pi
pred :: _ => _
pred = P.pred
print :: _ => _
print = P.print
product :: _ => _
product = P.product
properFraction :: _ => _
properFraction = P.properFraction
putChar :: _ => _
putChar = P.putChar
putStr :: _ => _
putStr = P.putStr
putStrLn :: _ => _
putStrLn = P.putStrLn
quot :: _ => _
quot = P.quot
quotRem :: _ => _
quotRem = P.quotRem
read :: _ => _
read = P.read
readFile :: _ => _
readFile = P.readFile
readIO :: _ => _
readIO = P.readIO
readList :: _ => _
readList = P.readList
readLn :: _ => _
readLn = P.readLn
readParen :: _ => _
readParen = P.readParen
reads :: _ => _
reads = P.reads
readsPrec :: _ => _
readsPrec = P.readsPrec
realToFrac :: _ => _
realToFrac = P.realToFrac
recip :: _ => _
recip = P.recip
rem :: _ => _
rem = P.rem
repeat :: _ => _
repeat = P.repeat
replicate :: _ => _
replicate = P.replicate
return :: _ => _
return = P.return
reverse :: _ => _
reverse = P.reverse
round :: _ => _
round = P.round
scaleFloat :: _ => _
scaleFloat = P.scaleFloat
scanl :: _ => _
scanl = P.scanl
scanl1 :: _ => _
scanl1 = P.scanl1
scanr :: _ => _
scanr = P.scanr
scanr1 :: _ => _
scanr1 = P.scanr1
seq :: _ => _
seq = P.seq
sequence :: _ => _
sequence = P.sequence
sequence_ :: _ => _
sequence_ = P.sequence_
show :: _ => _
show = P.show
showChar :: _ => _
showChar = P.showChar
showList :: _ => _
showList = P.showList
showParen :: _ => _
showParen = P.showParen
showString :: _ => _
showString = P.showString
shows :: _ => _
shows = P.shows
showsPrec :: _ => _
showsPrec = P.showsPrec
significand :: _ => _
significand = P.significand
signum :: _ => _
signum = P.signum
sin :: _ => _
sin = P.sin
sinh :: _ => _
sinh = P.sinh
snd :: _ => _
snd = P.snd
span :: _ => _
span = P.span
splitAt :: _ => _
splitAt = P.splitAt
sqrt :: _ => _
sqrt = P.sqrt
subtract :: _ => _
subtract = P.subtract
succ :: _ => _
succ = P.succ
sum :: _ => _
sum = P.sum
tail :: _ => _
tail = P.tail
take :: _ => _
take = P.take
takeWhile :: _ => _
takeWhile = P.takeWhile
tan :: _ => _
tan = P.tan
tanh :: _ => _
tanh = P.tanh
toEnum :: _ => _
toEnum = P.toEnum
toInteger :: _ => _
toInteger = P.toInteger
toRational :: _ => _
toRational = P.toRational
truncate :: _ => _
truncate = P.truncate
uncurry :: _ => _
uncurry = P.uncurry
undefined :: _ => _
undefined = P.undefined
unlines :: _ => _
unlines = P.unlines
until :: _ => _
until = P.until
unwords :: _ => _
unwords = P.unwords
unzip :: _ => _
unzip = P.unzip
unzip3 :: _ => _
unzip3 = P.unzip3
userError :: _ => _
userError = P.userError
words :: _ => _
words = P.words
writeFile :: _ => _
writeFile = P.writeFile
zip :: _ => _
zip = P.zip
zip3 :: _ => _
zip3 = P.zip3
zipWith :: _ => _
zipWith = P.zipWith
zipWith3 :: _ => _
zipWith3 = P.zipWith3
