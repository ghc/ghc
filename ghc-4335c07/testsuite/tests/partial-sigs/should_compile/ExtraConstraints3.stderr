TYPE SIGNATURES
  !! :: forall a. [a] -> Int -> a
  $ :: forall a b. (a -> b) -> a -> b
  $! :: forall a b. (a -> b) -> a -> b
  && :: Bool -> Bool -> Bool
  * :: forall a. Num a => a -> a -> a
  ** :: forall a. Floating a => a -> a -> a
  + :: forall a. Num a => a -> a -> a
  ++ :: forall a. [a] -> [a] -> [a]
  - :: forall a. Num a => a -> a -> a
  . :: forall b c a. (b -> c) -> (a -> b) -> a -> c
  / :: forall a. Fractional a => a -> a -> a
  /= :: forall a. Eq a => a -> a -> Bool
  < :: forall a. Ord a => a -> a -> Bool
  <= :: forall a. Ord a => a -> a -> Bool
  =<< ::
    forall (m :: * -> *) a b. Monad m => (a -> m b) -> m a -> m b
  == :: forall a. Eq a => a -> a -> Bool
  > :: forall a. Ord a => a -> a -> Bool
  >= :: forall a. Ord a => a -> a -> Bool
  >> :: forall (m :: * -> *) a b. Monad m => m a -> m b -> m b
  >>= ::
    forall (m :: * -> *) a b. Monad m => m a -> (a -> m b) -> m b
  ^ :: forall b a. (Integral b, Num a) => a -> b -> a
  ^^ :: forall a b. (Fractional a, Integral b) => a -> b -> a
  abs :: forall a. Num a => a -> a
  acos :: forall a. Floating a => a -> a
  acosh :: forall a. Floating a => a -> a
  all ::
    forall (t :: * -> *) a. P.Foldable t => (a -> Bool) -> t a -> Bool
  and :: forall (t :: * -> *). P.Foldable t => t Bool -> Bool
  any ::
    forall (t :: * -> *) a. P.Foldable t => (a -> Bool) -> t a -> Bool
  appendFile :: FilePath -> String -> IO ()
  asTypeOf :: forall a. a -> a -> a
  asin :: forall a. Floating a => a -> a
  asinh :: forall a. Floating a => a -> a
  atan :: forall a. Floating a => a -> a
  atan2 :: forall a. RealFloat a => a -> a -> a
  atanh :: forall a. Floating a => a -> a
  break :: forall a. (a -> Bool) -> [a] -> ([a], [a])
  ceiling :: forall a b. (RealFrac a, Integral b) => a -> b
  compare :: forall a. Ord a => a -> a -> Ordering
  concat :: forall (t :: * -> *) a. P.Foldable t => t [a] -> [a]
  concatMap ::
    forall (t :: * -> *) a b. P.Foldable t => (a -> [b]) -> t a -> [b]
  const :: forall a b. a -> b -> a
  cos :: forall a. Floating a => a -> a
  cosh :: forall a. Floating a => a -> a
  curry :: forall a b c. ((a, b) -> c) -> a -> b -> c
  cycle :: forall a. [a] -> [a]
  decodeFloat :: forall a. RealFloat a => a -> (Integer, Int)
  div :: forall a. Integral a => a -> a -> a
  divMod :: forall a. Integral a => a -> a -> (a, a)
  drop :: forall a. Int -> [a] -> [a]
  dropWhile :: forall a. (a -> Bool) -> [a] -> [a]
  either :: forall a c b. (a -> c) -> (b -> c) -> Either a b -> c
  elem ::
    forall (t :: * -> *) a. (P.Foldable t, Eq a) => a -> t a -> Bool
  encodeFloat :: forall a. RealFloat a => Integer -> Int -> a
  enumFrom :: forall a. Enum a => a -> [a]
  enumFromThen :: forall a. Enum a => a -> a -> [a]
  enumFromThenTo :: forall a. Enum a => a -> a -> a -> [a]
  enumFromTo :: forall a. Enum a => a -> a -> [a]
  error :: forall a. [Char] -> a
  even :: forall a. Integral a => a -> Bool
  exp :: forall a. Floating a => a -> a
  exponent :: forall a. RealFloat a => a -> Int
  fail :: forall (m :: * -> *) a. Monad m => String -> m a
  filter :: forall a. (a -> Bool) -> [a] -> [a]
  flip :: forall a b c. (a -> b -> c) -> b -> a -> c
  floatDigits :: forall a. RealFloat a => a -> Int
  floatRadix :: forall a. RealFloat a => a -> Integer
  floatRange :: forall a. RealFloat a => a -> (Int, Int)
  floor :: forall a b. (RealFrac a, Integral b) => a -> b
  fmap ::
    forall (f :: * -> *) a b. Functor f => (a -> b) -> f a -> f b
  foldl ::
    forall (t :: * -> *) b a.
    P.Foldable t =>
    (b -> a -> b) -> b -> t a -> b
  foldl1 ::
    forall (t :: * -> *) a. P.Foldable t => (a -> a -> a) -> t a -> a
  foldr ::
    forall (t :: * -> *) a b.
    P.Foldable t =>
    (a -> b -> b) -> b -> t a -> b
  foldr1 ::
    forall (t :: * -> *) a. P.Foldable t => (a -> a -> a) -> t a -> a
  fromEnum :: forall a. Enum a => a -> Int
  fromInteger :: forall a. Num a => Integer -> a
  fromIntegral :: forall a b. (Integral a, Num b) => a -> b
  fromRational :: forall a. Fractional a => Rational -> a
  fst :: forall a b. (a, b) -> a
  gcd :: forall a. Integral a => a -> a -> a
  getChar :: IO Char
  getContents :: IO String
  getLine :: IO String
  head :: forall a. [a] -> a
  id :: forall a. a -> a
  init :: forall a. [a] -> [a]
  interact :: (String -> String) -> IO ()
  ioError :: forall a. IOError -> IO a
  isDenormalized :: forall a. RealFloat a => a -> Bool
  isIEEE :: forall a. RealFloat a => a -> Bool
  isInfinite :: forall a. RealFloat a => a -> Bool
  isNaN :: forall a. RealFloat a => a -> Bool
  isNegativeZero :: forall a. RealFloat a => a -> Bool
  iterate :: forall a. (a -> a) -> a -> [a]
  last :: forall a. [a] -> a
  lcm :: forall a. Integral a => a -> a -> a
  length :: forall (t :: * -> *) a. P.Foldable t => t a -> Int
  lex :: ReadS String
  lines :: String -> [String]
  log :: forall a. Floating a => a -> a
  logBase :: forall a. Floating a => a -> a -> a
  lookup :: forall a b. Eq a => a -> [(a, b)] -> Maybe b
  map :: forall a b. (a -> b) -> [a] -> [b]
  mapM ::
    forall (t :: * -> *) (m :: * -> *) a b.
    (P.Traversable t, Monad m) =>
    (a -> m b) -> t a -> m (t b)
  mapM_ ::
    forall (t :: * -> *) (m :: * -> *) a b.
    (P.Foldable t, Monad m) =>
    (a -> m b) -> t a -> m ()
  max :: forall a. Ord a => a -> a -> a
  maxBound :: forall w. Bounded w => w
  maximum ::
    forall (t :: * -> *) a. (P.Foldable t, Ord a) => t a -> a
  maybe :: forall b a. b -> (a -> b) -> Maybe a -> b
  min :: forall a. Ord a => a -> a -> a
  minBound :: forall w. Bounded w => w
  minimum ::
    forall (t :: * -> *) a. (P.Foldable t, Ord a) => t a -> a
  mod :: forall a. Integral a => a -> a -> a
  negate :: forall a. Num a => a -> a
  not :: Bool -> Bool
  notElem ::
    forall (t :: * -> *) a. (P.Foldable t, Eq a) => a -> t a -> Bool
  null :: forall (t :: * -> *) a. P.Foldable t => t a -> Bool
  odd :: forall a. Integral a => a -> Bool
  or :: forall (t :: * -> *). P.Foldable t => t Bool -> Bool
  otherwise :: Bool
  pi :: forall w. Floating w => w
  pred :: forall a. Enum a => a -> a
  print :: forall a. Show a => a -> IO ()
  product ::
    forall (t :: * -> *) a. (P.Foldable t, Num a) => t a -> a
  properFraction ::
    forall a b. (RealFrac a, Integral b) => a -> (b, a)
  putChar :: Char -> IO ()
  putStr :: String -> IO ()
  putStrLn :: String -> IO ()
  quot :: forall a. Integral a => a -> a -> a
  quotRem :: forall a. Integral a => a -> a -> (a, a)
  read :: forall a. Read a => String -> a
  readFile :: FilePath -> IO String
  readIO :: forall a. Read a => String -> IO a
  readList :: forall a. Read a => ReadS [a]
  readLn :: forall a. Read a => IO a
  readParen :: forall a. Bool -> ReadS a -> ReadS a
  reads :: forall a. Read a => ReadS a
  readsPrec :: forall a. Read a => Int -> ReadS a
  realToFrac :: forall a b. (Real a, Fractional b) => a -> b
  recip :: forall a. Fractional a => a -> a
  rem :: forall a. Integral a => a -> a -> a
  repeat :: forall a. a -> [a]
  replicate :: forall a. Int -> a -> [a]
  return :: forall (m :: * -> *) a. Monad m => a -> m a
  reverse :: forall a. [a] -> [a]
  round :: forall a b. (RealFrac a, Integral b) => a -> b
  scaleFloat :: forall a. RealFloat a => Int -> a -> a
  scanl :: forall b a. (b -> a -> b) -> b -> [a] -> [b]
  scanl1 :: forall a. (a -> a -> a) -> [a] -> [a]
  scanr :: forall a b. (a -> b -> b) -> b -> [a] -> [b]
  scanr1 :: forall a. (a -> a -> a) -> [a] -> [a]
  seq :: forall a b. a -> b -> b
  sequence ::
    forall (t :: * -> *) (m :: * -> *) a.
    (P.Traversable t, Monad m) =>
    t (m a) -> m (t a)
  sequence_ ::
    forall (t :: * -> *) (m :: * -> *) a.
    (P.Foldable t, Monad m) =>
    t (m a) -> m ()
  show :: forall a. Show a => a -> String
  showChar :: Char -> ShowS
  showList :: forall a. Show a => [a] -> ShowS
  showParen :: Bool -> ShowS -> ShowS
  showString :: String -> ShowS
  shows :: forall a. Show a => a -> ShowS
  showsPrec :: forall a. Show a => Int -> a -> ShowS
  significand :: forall a. RealFloat a => a -> a
  signum :: forall a. Num a => a -> a
  sin :: forall a. Floating a => a -> a
  sinh :: forall a. Floating a => a -> a
  snd :: forall a b. (a, b) -> b
  span :: forall a. (a -> Bool) -> [a] -> ([a], [a])
  splitAt :: forall a. Int -> [a] -> ([a], [a])
  sqrt :: forall a. Floating a => a -> a
  subtract :: forall a. Num a => a -> a -> a
  succ :: forall a. Enum a => a -> a
  sum :: forall (t :: * -> *) a. (P.Foldable t, Num a) => t a -> a
  tail :: forall a. [a] -> [a]
  take :: forall a. Int -> [a] -> [a]
  takeWhile :: forall a. (a -> Bool) -> [a] -> [a]
  tan :: forall a. Floating a => a -> a
  tanh :: forall a. Floating a => a -> a
  toEnum :: forall a. Enum a => Int -> a
  toInteger :: forall a. Integral a => a -> Integer
  toRational :: forall a. Real a => a -> Rational
  truncate :: forall a b. (RealFrac a, Integral b) => a -> b
  uncurry :: forall a b c. (a -> b -> c) -> (a, b) -> c
  undefined :: forall w. w
  unlines :: [String] -> String
  until :: forall a. (a -> Bool) -> (a -> a) -> a -> a
  unwords :: [String] -> String
  unzip :: forall a b. [(a, b)] -> ([a], [b])
  unzip3 :: forall a b c. [(a, b, c)] -> ([a], [b], [c])
  userError :: String -> IOError
  words :: String -> [String]
  writeFile :: FilePath -> String -> IO ()
  zip :: forall a b. [a] -> [b] -> [(a, b)]
  zip3 :: forall a b c. [a] -> [b] -> [c] -> [(a, b, c)]
  zipWith :: forall a b c. (a -> b -> c) -> [a] -> [b] -> [c]
  zipWith3 ::
    forall a b c d. (a -> b -> c -> d) -> [a] -> [b] -> [c] -> [d]
  || :: Bool -> Bool -> Bool
TYPE CONSTRUCTORS
COERCION AXIOMS
Dependent modules: []
Dependent packages: [base-4.11.0.0, ghc-prim-0.5.2.0,
                     integer-gmp-1.0.1.0]
